### DISCUSSION TITLE
Question re. Actuator & Prometheus

### DISCUSSION THREAD

**Discussion by stwomack** [promregator/promregator#185] [Ideas]
Would promregator be able to support allowing teams to scrap actuator endpoints, in order to shuttle the scraped metrics to off-platform prometheus?
/cc @xyloman

**Comment by checketts** [comment]
@stwomack Are you referring to the Prometheus actuator endpoint? If so, that is exactly what Promregator is intended for. Can you add some more details in what you are doing and what isn't working?

If you are asking about non-prometheus actuator endpoints (like /health and /info) then that is out of the scope of Promregator

**Comment by xyloman** [comment]
Correct this would be scraping spring boot actuator endpoints preferably with authentication against said endpoints. This would avoid the need to do container to container networking just to fetch metrics on a different port which is done today to prevent access since metrics register does not have authentication. We also would be interested if promaggator could be registered with the metrics register. This would allow the discovery and authentication capabilities of promaggator to marry up with metrics register

**Comment by checketts** [comment]
Ah I see. You are referring to the feature of exposing actuator endpoints on a different port. I don't believe that is currently supported, but I think that is an excellent feature that would align really well with Promregator. (However I suspect some aspect of container networking might be required since Promregator can't bypass the CloudFoundry networking)

> We also would be interested if promaggator could be registered with the metrics register. This would allow the discovery and authentication capabilities of promaggator to marry up with metrics register

Can you tell me more about the 'metrics register'? Perhaps a link to the documentation explaining what you are referring to. Is that a Cloud Foundry feature?

**Comment by xyloman** [comment]
[Metrics registar](https://docs.pivotal.io/platform/application-service/2-8/metric-registrar/using.html) is a component that allows a developer to bind a service to their application that tells the platform what endpoint to scrape.  However, it lacks the ability today to do authentication.  It also can not perform the scraping over a c2c only networking path.  Promregator allows for discovery of applications running on the platform and leverages conventions to attempt to scrap actuator endpoints.  Metric registrar is a capability where the developer has to choose to bind the service to the application which in turn registers the application to be scraped.  Once registered the metrics are then sent to the [metrics store](https://github.com/cloudfoundry/metric-store-release) which inherits RBAC authorization controls from CF so that a space developer in a given space can the right to query their metrics but they do not have the rights to query metrics from spaces in which they are not a space developer.  This can be leveraged with Grafana for charting purposes. 

So I have been curious about the possibility of merging these two approaches together. Where in which we deploy promregator as an application running on cf and generate it cf credentials so that it can discover all of the applications running on the platform.   Those applications would then use an agreed upon authentication mechanic to protect their actuator endpoints.  Proregator would scrape all of the registered applications and then register its endpoint with the metrics registar so that the metrics collected can end up in the metrics store.  If a team desires to shuttle these metrics to a different prometheus they can also do that as well by scraping from the promregator.

**Comment by ScottAntle** [comment]
Keeping on the radar.

**Comment by eaglerainbow** [comment]
Reading through this discussion here, and in particular the readme of https://github.com/cloudfoundry/metric-store-release, I find it interesting. However, I have the impression that the two concepts are built based on different assumptions: Promregator's idea is that you do not need any kind of support from the platform administrator to be enable your applications to scrape metrics (mainly as, when Promregator was built initially, all the community folks were reluctant for such kind of support). The metrics store/registrar principle is based on the assumption that the CF administrators will support you by deploying components on BOSH level. 
To that end, the two approaches appear to be mutual exclusive to me. 

Or did I miss a major aspect in this discussion here?

**Comment by xyloman** [comment]
For me the intersection is metrics registrar is the means to get the metrics in the metrics store. We like the approach that Promregator takes around discovery and collection. However, instead of sending the metrics to prometheus we are interested in sending them to metrics store. Ny understanding is the only way to accomplish this is through the metrics registrar hence this inquiry.

**Comment by eaglerainbow** [comment]
So, do I get you right that your idea is to use metrics registrar to make use of Promregator for scraping (instead of Loggregator)?
This feels somewhat strange to me...

**Comment by eaglerainbow** [comment]
BTW: It might be worthwhile to also read #61 - it discusses the implications of the approach to replace the discovery mechanism of Promregator with a registry-based one (there: service-broker-/OSB-based). There are several cons on that.
